js/web/lib/wasm/jsep/webgpu/ops/reduce.ts

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import {DataType} from '../../../wasm-common';
+import {TensorView} from '../../tensor';
+import {ShapeUtil} from '../../util';
+import {AttributeWithCacheKey, createAttributeWithCacheKey} from '../attribute-with-cache-key';
+import {ComputeContext, GpuDataType, ProgramInfo, ProgramInfoLoader, ProgramMetadata} from '../types';
+
+import {createIndicesHelper, ShaderHelper} from './common';
+
+const validateInputs = (inputs: readonly TensorView[]): void => {
+  if (!inputs || inputs.length == 0 || inputs.length > 2) {
+    throw new Error('Reduce op requires 1 or 2 inputs.');
+  }
+
+  if (inputs[0].dataType !== DataType.float) {
+    throw new Error('Invalid input type.');
+  }
+};
+
+export interface ReduceAttributes extends AttributeWithCacheKey {
+  axes: number[];
+  keepDims: boolean;
+  noopWithEmptyAxes: boolean;
+}
+
+type ReduceOp = (inputs: readonly TensorView[], axes: number[]) => string[];
+
+const createReduceProgramInfo =
+    (metadata: ProgramMetadata, inputs: readonly TensorView[], attributes: ReduceAttributes,
+     reduceOp: ReduceOp): ProgramInfo => {
+      const outputShape: number[] = [];
+      const inputShape = inputs[0].dims;
+
+      const idxCopy: string[] = [];  // copy output indexes to input indexes
+
+      const axes = ShapeUtil.normalizeAxes(attributes.axes, inputs[0].dims.length);
+      const ops = reduceOp(inputs, axes);
+      const inputIndicesHelper = createIndicesHelper('input', inputShape);
+      const initInputIdx = (ops[1] === '') ? '' : `let inputIdx = ${inputIndicesHelper.i2oExpression('inputIndices')};`;
+      let reduceOps = `
+          let inputIdx = ${inputIndicesHelper.i2oExpression('inputIndices')};
+          ${ops[2]};`;
+
+      for (let k = 0; k < inputs[0].dims.length; k++) {
+        // if this axis is reduced
+        if (axes.indexOf(k) >= 0 || axes.length === 0) {
+          if (attributes.keepDims === true) {
```

Comment:
`=== true` is no redundant

Reply:
Do you mean that it is redundant? I am just being explicit and I am checking a boolean value against a boolean const.

Reply:
Is this wrong?

Reply:
this is not wrong. however, I think it's considered a better way to check `if (attributes.keepDims)` than `if (attributes.keepDims === true)` given that `attributes.keepDims` is boolean type.